lib/mix/lib/mix/scm/git.ex

```diff
@@ -104,6 +108,15 @@ defmodule Mix.SCM.Git do
 
   ## Helpers
 
+  # TODO: make it raise (at v2.0?)
+  defp validate_git_options(opts) do
+    keys = Keyword.keys(opts)
+    if Enum.count(keys, fn(key) -> Enum.member?([:branch, :ref, :tag], key) end) > 1 do
+      Mix.shell.error("warning: you should specify only one of branch, ref or tag, and only once")
```

Comment:
Sweet! Btw, do you have the app name in `opts`? If not, we should tell the git/github info, so the warning is clear. Something like:

    Mix.shell.error("warning: you should specify only one of branch, ref or tag, and only once. Error on git dependency #{git_or_app}")